Adding CONTRIBUTING.md file (#2)
redhat-beyond/MeetBalls: CONTRIBUTING.md

Reviewer: I'd also suggest committing to some commit message convention (pick one which would be reasonable to follow in the context of this course).
See an example here: https://cbea.ms/git-commit/#seven-rules

Author: I'll research more about it, and get an answer by tomorrow. Thank you for the reference.

Author: Would you recommend using [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) in our project? Very structural, very easy to understand.
My fear is that it's a bit too much, and maybe we should use the Seven Rules convention (the link you've attached), which is easier to understand and very simple.

Reviewer: I'd prefer those seven rules in this case. 

In Red Hat, we're using conventional commits very often. I like it. The structure of the commit header might cause a little confusion initially.

I like the link that I shared because it summarizes some aspects of the content of the commit message rather than the technical structure, and I think this is more important at this stage.

# Contribution Guide

## Getting Started

### Prerequisites
Please make sure you have the following programs installed:
- [Python 3.9](https://www.python.org/downloads/)
- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
- [Virtual Box](https://www.virtualbox.org/wiki/Downloads)
- [Vagrant](https://www.vagrantup.com/downloads)

IMPORTANT:

> - Note that Vagrant may not work on some machines (such as Apple M1-based machines), so please make sure that you can find a suitable solution.

Author: I thought that VirtualBox was not compatible with Apple M1 processors, but it turns out that in the past few months they have updated their software and now it's fully functioning. Next time I'll double check, and not trust my memory.
Thank you for the comment.

> - Note that Vagrant may not work on some machines (such as Apple M1-based machines), so please make sure that you can find a suitable solution.

Reviewer: I meant to say that it wasn't an issue with vagrant's compatibility, but with virtualbox's.

### Setting up
1. Fork the repository.
2. Clone the repository locally.
3. Create a new branch.
4. Open any terminal and navigate to the project directory.
5. Run the `vagrant up` command.

After the dependencies were installed and the VM initialized, you can proceed working on the project code.

> ## Development Process

Author: I've forgot adding these comments. Right now I've uploaded a new PR, containing these recommendations. Thank you.

> ## Development Process

Reviewer 2: It seems that you have responded to all of the reviews except for the one suggesting that pull requests should be made as small as possible. While you did mention in your response that the description of the PR should be concise, it may also be helpful to consider the amount of files and code logic in the PR to make it as minimal as possible

> ## Development Process

Author: Thank you for pointing that out. I really missed that part. It's now fixed and an updated version should be up soon.

### Issues
- The title must be clear, and describe what the issue is about.
- The description of the issue must include a clear explanation about the reasons for doing the work.
- Status Labels(Optional):
  - Status/New - An issue that was just created, in addition unlabeled issues are considered new.
  - Status/Backlog - A issue that was discussed and accepted for work.
  - Status/In progress - An issue that is being worked on.

> ### Pull Requests

Reviewer 2: Im missing info about reviewers and assignee labeling, this will ensure efficient and effective collaboration among contributor

> ### Pull Requests

Author: Thank you for your comment. I can't understand the part of "Im missing info about reviewers and assignee labeling". As we can see at "Labels" tab (https://github.com/redhat-beyond/MeetBalls/labels), all the statuses are described there, except the parts of "Status/New", "Status/Backlog", "Status/In progress", and that's why I've described them in this file. Can you please explain this part to me?

> ### Pull Requests

Reviewer 2: i understand the confusion,
I meant you can add information about [Request Reviewers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) and [Assign User To PR and Issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users) that will benefit the work flow

> ### Pull Requests

Author: Thank you for your comment. Will be added.

- **The Pull Request should not break any of the existing functionality**
- Each Pull Request should focus on single repository principle.
- <ins>Description-</ins> the PR description should explain what changes have been made, and why. It is recommended to include a link to the relevant `issue`
- Each PR should include no more than 5 `commits`

> - Every PR will be tested by:

Reviewer 2: The pull request will not undergo review until all tests have passed

    1. A flake8 workflow to make sure the PR conforms to proper Python syntax and coding standards.
    2. Automatically checked by a CI system.
- Each pull request requires the approval of at least <ins>2 team members</ins> before merging.
 
### Reviewing Pull Requests
- Each PR requires the approval of at least two team members. 

> - Examine the PR carefully, and with respect to what it is supposed to achieve.

Author: This part is related to "Reviewing PRs", correct me if I'm wrong, but I think that your comment is related to "Issuing PRs".
This subject ("Issuing PRs") was not presented in my current PR, but would be added in the updated one by tomorrow.
And I added your reference (test coverage) to "Issuing PRs".
Thank you.

> - Examine the PR carefully, and with respect to what it is supposed to achieve.

Reviewer: It's something for the reviewer to keep in mind as well, so it's good to have it here as well.

# Contribution Guide

## Getting Started

### Prerequisites
Please make sure you have the following programs installed:
- [Python 3.9](https://www.python.org/downloads/)
- [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
- [Virtual Box](https://www.virtualbox.org/wiki/Downloads)
- [Vagrant](https://www.vagrantup.com/downloads)

### Setting up
1. Clone your fork.
2. Create a new branch.
3. Open any terminal and navigate to the project directory.
4. Run the `vagrant up` command.

After the dependencies were installed and the VM initialized, you can proceed working on the project code.

## Development Process

> ### Commits

Reviewer: ```suggestion
### Commit Messages
```

- Separate subject from body with a blank line.
- Limit the subject line to 50 characters.
- Capitalize the subject line.
- Do not end subject line with a period.
- Use the imperative mood in the subject line.
- Wrap the body at 72 characters.
- Use the body to explain what and why vs. how.

For more information and examples, please refer to [this guide](https://cbea.ms/git-commit/#seven-rules).

### Issues
- The title must be clear, and describe what the issue is about.
- The description of the issue must include a clear explanation about the reasons for doing the work.
- Status Labels(Optional):
  - Status/New - An issue that was just created, in addition unlabeled issues are considered new.
  - Status/Backlog - A issue that was discussed and accepted for work.
  - Status/In progress - An issue that is being worked on.

### Pull Requests
- **The Pull Request should not break any of the existing functionality**

> - <ins>Description-</ins> the PR description should explain what changes have been made, and why. Incase it's related to a specific `issue` - it is mandatory to include a link to it.

Reviewer: You're using underline and dash here here, but on other similar occasions you no underline and colon.

> - <ins>Description-</ins> the PR description should explain what changes have been made, and why. Incase it's related to a specific `issue` - it is mandatory to include a link to it.
> Please refer to this [guide](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) of how to link a PR to an issue.
> - Each PR should include no more than 5 `commits`
> - The pull request will not undergo review until all tests have passed:
>     1. A [flake8](https://www.flake8rules.com/) workflow to make sure the PR conforms to proper Python syntax and coding standards.
>     2. Automatically checked by a CI system.
> - Each pull request requires the approval of at least <ins>2 team members</ins> before merging.
> - Each PR should [request reviewers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review). Reviewers should be people who are knowledgeable about the code you're modifying or related to the specific issue.
> Please refer to this [guide](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users) in order to assign users to PRs and issues.

Reviewer: I'd break those long lines (here and on other places) to make it easier to read. You can arbitrarily break lines in markdown files and it won't create a newline on the rendered output (as long as you don't leave an empty line).

Basically, if you have links, then you won't break the link in the middle (but you can leave long links on lines of their own), but for all other content, you can break it at a common number that fits in an editor (89 characters is a nice number that typically fits having 2 documents side-by-side on the same monitor with reasonable text size - but it's up to you).

I'd suggest breaking 
```suggestion
- <ins>Description-</ins> the PR description should explain what changes have been made, and why.
In case it's related to a specific `issue` - it is mandatory to include a link to it. Please refer to this
 [guide](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
 for how to link a PR to an issue.
- Each PR should include no more than 5 `commits`
- The pull request will not undergo review until all tests have passed:
    1. A [flake8](https://www.flake8rules.com/) workflow to make sure the PR conforms to proper Python syntax
        and coding standards.
    2. Automatically checked by a CI system.
- Each pull request requires the approval of at least <ins>2 team members</ins> before merging.
- Each PR should
   [request reviewers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review).
   Reviewers should be people who are knowledgeable about the code you're modifying or related to the
   specific issue.
Please refer to this
[guide](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users)
 in order to assign users to PRs and issues.
```

### Issuing Pull Request
- Title: Make sure the title is short and descriptive enough.
- Body: Do not assume full familiarity with the story or work that has been done.
Imagine that the code reviewer is arriving in your team today without knowing what is going on, and even so, they should be able to understand the changes. 
- Comments: Add comments to the pull request itself. Comment on portions of the code that you think might need further explanation or are complex.

> You can also tag people at specific portions of the pull request if you think want feedback from them at the point, or you believe that part concerns them.

Reviewer: ```suggestion
You can also tag people at specific portions of the pull request if you want feedback from them at that point, or you believe that that part concerns them.
```

## Development Process

### Commit Messages
- Separate subject from body with a blank line.
- Limit the subject line to 50 characters.
- Capitalize the subject line.
- Do not end subject line with a period.
- Use the imperative mood in the subject line.
- Wrap the body at 72 characters.
- Use the body to explain what and why vs. how.

For more information and examples, please refer to [this guide](https://cbea.ms/git-commit/#seven-rules).

### Issues
- The title must be clear, and describe what the issue is about. 
- The description of the issue must include a clear explanation about the reasons for doing the work. Also, it should be as small as possible.
- Status Labels(Optional):
  - Status/New - An issue that was just created, in addition unlabeled issues are considered new.
  - Status/Backlog - A issue that was discussed and accepted for work.
  - Status/In progress - An issue that is being worked on.

### Pull Requests
- The Pull Request should not break any of the existing functionality
- Description: the PR description should be as small as possible, and explain what changes have been made, and why.

> In case it's related to a specific `issue` - it is mandatory to include a link to it.

Reviewer 2: To submit a PR, it is mandatory to link it to a relevant issue. This means that if there is no existing issue that relates to your PR, you must create one before you can submit your request.
please make it more implicit in the Pull Requests section

Author: I was thinking that "Issue" section is related mainly to report bugs in our code.
That's why I haven't included the part of "This means that if there is no existing issue that relates to your PR, you must create one before you can submit your request.". Because what happens when the contributor creates a new feature? New features would be presented in the issues section?
Thank you for your comment.

Reviewer 2: Its not focused on bugs only , it should also be used to discuss new features or improvements to existing functionality. see more [About issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/about-issues)

Author: Thank you for your comment. Now it all makes sense. I've updated the file, and a new commit will appear soon.
